packages/core/src/new-api/internal/execution/execution-engine.ts

```diff
@@ -0,0 +1,219 @@
+
+import { executionStateReducer } from "./executionStateReducer";
+import { isExecutionResult, isOnChainAction, isOnchainResult } from "./guards";
+
+type ExecutionBatch = Future[];
+
+export class ExecutionEngine {
+  public async execute(state: ExecutionEngineState): Promise<DeploymentResult> {
+    const { batches, module } = state;
+
+    const futures = module.getFutures();
+
+    for (const batch of batches) {
+      // TODO: consider changing batcher to return futures rather than ids
+      const executionBatch = batch.map((futureId) =>
+        this._lookupFuture(futures, futureId)
+      );
+
+      const batchResult = await this._executeBatch(executionBatch, state);
+
+      if (batchResult.some((b) => b.type === "execution-failure")) {
+        return { status: "failed" };
+      }
+
+      if (batchResult.some((b) => b.type === "execution-hold")) {
+        return { status: "hold" };
+      }
+
+      if (batchResult.every((b) => b.type !== "execution-success")) {
+        throw new IgnitionError("Unexpected state");
+      }
+    }
+
+    return {
+      status: "success",
+      contracts: this._resolveDeployedContractsFrom(state),
+    };
+  }
+
+  private async _executeBatch(
+    batch: ExecutionBatch,
+    state: ExecutionEngineState
+  ) {
+    return Promise.all(
+      batch.map((entry) => this._executeBatchEntry(entry, state))
+    );
+  }
+
+  private async _executeBatchEntry(
+    future: Future,
+    state: ExecutionEngineState
+  ): Promise<ExecutionResult> {
+    let current: JournalableMessage = this._startOrRestartFor(future, state);
+    await this._apply(state, current);
+
+    const context = {
+      executionState: state.executionStateMap[future.id],
+    };
+
+    const exectionStrategy = state.strategy.executeStrategy(context);
+
+    while (!isExecutionResult(current)) {
+      context.executionState = state.executionStateMap[future.id];
+
+      if (isOnChainAction(current)) {
+        current = await state.transactionService.onchain(current);
+      } else if (isOnchainResult(current)) {
+        current = (await exectionStrategy.next(current)).value;
+      } else {
+        current = (await exectionStrategy.next(null)).value;
+      }
+
+      await this._apply(state, current);
+    }
+
+    return current;
+  }
+
+  private _startOrRestartFor(
+    future: Future,
+    state: ExecutionEngineState
+  ): FutureStart | FutureRestart {
+    const executionState = state.executionStateMap[future.id];
+
+    if (executionState === undefined) {
+      return this._initCommandFor(future, state);
+    }
+
+    return { type: "execution-restart", futureId: future.id };
+  }
+
+  private async _apply(
+    state: ExecutionEngineState,
+    message: JournalableMessage
+  ): Promise<void> {
+    await state.journal.record(message);
+    state.executionStateMap = executionStateReducer(
+      state.executionStateMap,
+      message
+    );
+  }
+
+  private _resolveAddress(
```

Comment:
Thinking out loud: this feels a bit duplicated wrt the reconciliation pr, maybe we can avoid the repetition (not convinced it's worth it though)

Reply:
I just attempted to pull it out. But the resolveAddress logic is subtely different from reconciliation.

Specifically in reconciliation, we treat `from` set to undefined as a valid value. Here we replace undefined with `Accounts[0]`.

Because accounts can change between runs, should reconciliation keep the current behaviour of allowing undefined, or should it substitute in `accounts[0]`?

Reply:
I think once we started executing a future with a certain from that should be considered fixed. If during a restart the future doesn't specify a `from`, and resolving `accounts[0]` leads to a different account, that should be a reconciliation error.

Reply:
Cool. That will allow us to share the logic. I will add a todo to pull that out.